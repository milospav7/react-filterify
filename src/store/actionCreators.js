import {
  OVERRIDE_STATE,
  RESET_ALL_FILTERS,
  RESET_POPERTY_FILTERS,
  UPDATE_FILTER,
  UPDATE_FUNCTION_FILTER,
  UPDATE_MULTIPLE_FUNCTION_FILTERS,
  UPDATE_NAVIGATION_PROPERTY_FILTER,
} from "./actionTypes";

/**
 * Updates propertyFilters map inside filter instance
 * @param {*} id redux filter Id - must be initialized inside gridFilters.js reducer
 * @param {*} filteringProperty name of the property/field used for filtering. If contains ';' char then it will be processed as multi filter - filtering more than one field with one common filter(value). Example: 'Width;Height'
 * @param {*} filteringValue value used for filtering - both number/string and array of number/string are supported
 * @param {*} operator operator used for field filtering - contains, eq, neq etc..
 * @param {*} logic operator used as logic operator between multiple values..
 */
export const updatePropertyFilter = (
  id,
  filteringProperty,
  filteringValue,
  operator = "eq",
  logic = "or",
  allowNullValue = false
) => ({
  type: UPDATE_FILTER,
  id,
  filteringProperty,
  filteringValue,
  operator,
  logic,
  allowNullValue,
});

/**
 * Resets all in filter instance, reverts to initial state
 * @param {*} id redux filter Id - must be initialized inside gridFilters.js reducer
 */
export const resetAllFilters = (id, keepTreeViewFilter = false) => ({
  type: RESET_ALL_FILTERS,
  id,
  keepTreeViewFilter,
});

/**
 * Resets all in filter instance, reverts to initial state
 * @param {*} id redux filter Id - must be initialized inside gridFilters.js reducer
 * @param {*} filteringProperties collections of filter names that should be reseted
 */
export const resetPropertyFiltersByNames = (id, filteringProperties) => ({
  type: RESET_POPERTY_FILTERS,
  id,
  filteringProperties,
});

/**
 *	Updates navigationPropetyFilters map inside filter instance
 * @param {*} id redux filter Id - must be initialized inside gridFilters.js reducer
 * @param {*} navigationProperty name of collection based property that we want to filter with lambda expression
 * @param {*} filteringProperty name of the property/field of navigation entity, used for filtering inside lambda expression
 * @param {*} filteringValue value used for filtering - both number/string and array of number/string are supported
 * @param {*} generatedExpression use only when you want to pass functions inside lambda expression(when filteringValue argument is not good enough) - for example if we want to include 'inCachedView(s/SubstationID, 21fdb7bb-39db-435d-af9a-60d695303ad1, true)' in navigation property lambda, then we need to pass this 'inCachedView(s/SubstationID, 21fdb7bb-39db-435d-af9a-60d695303ad1, true)' using generatedExpression argument
 * @param {*} navigationPropertyIsNested indicates whether provided navigation property name represents nav property inside nav property(important when constructing query string, because if it is nested it means that two lambdas will be created for this filter - one inside another). Example: navigationProperty: 'Users/Books', this is nested and will generate smth like '...Users/any(u: Books/any(b: andHereComesgeneratedExpressionOrfilteringValuesWithOperators))..' At the moment we support maximum of 2 level nesting.
 * @example1 updateNavigationPropertyFilter(12, 'RelatedSubstations', 'OperationalRegion', 'OpZone 1', null) - for final filter result(after using gridFilter reducer helper methods): RelatedSubstations/any(s: ( s
 * @example2 updateNavigationPropertyFilter(12, 'RelatedSubstations', 'OperationalRegion', '21fdb7bb-39db-435d-af9a-60d695303ad1', 'inCachedView(s/SubstationID, 21fdb7bb-39db-435d-af9a-60d695303ad1, true)') for final filter result(after using gridFilter reducer helper methods): RelatedSubstations/any(s: inCachedView(s/SubstationID, 21fdb7bb-39db-435d-af9a-60d695303ad , true)), /OperationalRegion eq 'OpZone 1')
 * @note There is one limitation when using generatedExpression argument. If you need lambda input parameter inside passed expression, use 's' for lambda input-paramter, example: 'isNotEmpty(s.Name)',
 */
export const updateNavigationPropertyFilter = (
  id,
  navigationProperty,
  filteringProperty,
  filteringValue,
  generatedExpression = null,
  navigationPropertyIsNested = false
) => ({
  type: UPDATE_NAVIGATION_PROPERTY_FILTER,
  id,
  navigationProperty,
  filteringProperty,
  filteringValue,
  generatedExpression,
  navigationPropertyIsNested,
});

/**
 * Updates functionFilters array inside filter instance
 * @param {*} id redux filter Id - must be initialized inside gridFilters.js reducer
 * @param {*} filteringProperty unique identifier of function filter string. Should be field name
 * @param {*} filterQueryString function filter that needs to be included in query string - non standard functions
 * @param {*} values this can be used as a container for values that are provided inside function query string, as a separete object for putting raw values only. Format/structure is not strict, and should be constrained only by the way of retreiving these values back in caller component - for example when we need to read these values on component mount we can use this object from redux(because we can't read them from raw query string)
 * @example filterQueryString: 'insideTimeFrame(PlannedStartTime, PlannedEndTime, 2020-06-30T22:00:00.000Z, null, false, false)'
 */
export const updateFunctionFilter = (
  id,
  filteringProperty,
  filterQueryString,
  values = null
) => ({
  type: UPDATE_FUNCTION_FILTER,
  id,
  filteringProperty,
  filterQueryString,
  values,
});

export const updateMultipleFunctionFilters = (id, filtersToUpdate) => ({
  type: UPDATE_MULTIPLE_FUNCTION_FILTERS,
  id,
  filtersToUpdate,
});

export const overrideFilterState = (id, filterState) => ({
  type: OVERRIDE_STATE,
  id,
  filterState,
});
