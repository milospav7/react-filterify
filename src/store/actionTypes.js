export const FILTER_UPDATE_FILTER = "[SHARED] FILTER_UPDATE_FILTER";
export const FILTER_RESET_ALL_FILTERS = "[SHARED] FILTER_RESET_ALL_FILTERS";
export const FILTER_RESET_POPERTY_FILTERS =
  "[SHARED] FILTER_RESET_POPERTY_FILTERS";
export const FILTER_UPDATE_NAVIGATION_PROPERTY_FILTER =
  "[SHARED] FILTER_UPDATE_NAVIGATION_PROPERTY_FILTER";
export const FILTER_UPDATE_FUNCTION_FILTER =
  "[SHARED] FILTER_UPDATE_FUNCTION_FILTER";
export const FILTER_UPDATE_MULTIPLE_FUNCTION_FILTERS =
  "[SHARED] FILTER_UPDATE_MULTIPLE_FUNCTION_FILTERS";
export const FILTER_OVERRIDE_FILTER_STATE =
  "[SHARED] FILTER_OVERRIDE_FILTER_STATE";
