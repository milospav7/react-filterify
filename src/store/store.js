import { createStore, combineReducers } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import { configureFilterfyReducer } from "./orchestratorReducer";

export const CONTAINER_IDS = {
  C1: "C1",
  C2_Test: "C2_Test",
};

const composedEnhancers = composeWithDevTools();
const filterifyFilters = configureFilterfyReducer([
  {
    id: CONTAINER_IDS.C1,
    saveToLocalStorage: true,
    styleSchema: {
      highlightWhenInUse: true,
      labelColor: "red",
      labelFontSize: "13px",
    },
  },
  {
    id: CONTAINER_IDS.C2_Test,
    saveToLocalStorage: false,
  },
]);

const reducers = combineReducers({
  filterifyFilters,
});

const store = createStore(reducers, composedEnhancers);

export default store;
