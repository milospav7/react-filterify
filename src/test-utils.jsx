// test-utils.jsx
import { render as rtlRender } from "@testing-library/react";
import { Provider } from "react-redux";
import appStore from "./store/store";

function render(
  ui,
  { preloadedState, store = appStore, ...renderOptions } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from "@testing-library/react";
// override render method
export { render };
