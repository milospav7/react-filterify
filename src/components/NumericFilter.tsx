/* eslint-disable react-hooks/exhaustive-deps */
import React, { useState, useMemo, useRef } from "react";
import { useDispatch } from "react-redux";
import {
  InputGroup,
  InputGroupAddon,
  DropdownToggle,
  DropdownMenu,
  DropdownItem,
  ButtonDropdown,
  InputGroupText,
  Input,
} from "reactstrap";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faCaretDown,
  faEquals,
  faGreaterThan,
  faGreaterThanEqual,
  faLessThan,
  faLessThanEqual,
  faNotEqual,
} from "@fortawesome/free-solid-svg-icons";
import { ValueTypedObject } from "../store/types";
import { useFilterActions, useFilterState } from "./hooks";
import { updatePropertyFilter } from "../store/actionCreators";
import { DebouncedInputField } from "./DebouncedInputField";
import { BaseFilterProps } from "../store/interfaces";
import FilterDecorator from "./FilterDecorator";

const operatorSymbols: ValueTypedObject<string> = {
  eq: "eq",
  ne: "ne",
  gt: "gt",
  ge: "ge",
  lt: "lt",
  le: "le",
};

const faIconByOperator: ValueTypedObject<any> = {
  eq: faEquals,
  ne: faNotEqual,
  gt: faGreaterThan,
  ge: faGreaterThanEqual,
  lt: faLessThan,
  le: faLessThanEqual,
};

interface IProps extends BaseFilterProps {
  placeholder?: string;
  useDecimal?: boolean;
}

const NumericFilter: React.FC<IProps> = ({
  containerId,
  filteringProperty,
  navigationProperty,
  useDecimal,
  placeholder,
  className,
  withAssociatedLabel,
  labelClassName,
  label,
  style,
}) => {
  const [dropdownOpen, setOpen] = useState(false);
  const toggle = () => setOpen(!dropdownOpen);
  const dispatcher = useDispatch();
  const inputRef = useRef<Input>(null);

  const { updateFilter } = useFilterActions(
    containerId,
    filteringProperty,
    navigationProperty
  );
  const { filterValue, filterOperator } = useFilterState(
    containerId,
    filteringProperty,
    navigationProperty
  );
  const operator = filterOperator ?? operatorSymbols.eq;

  const updateTargetFilter = (value: string | null) => {
    if (value) {
      const parsed = useDecimal ? parseFloat(value) : parseInt(value, 10);
      updateFilter(parsed, { operator });
    } else updateFilter(value, { operator });
  };

  const updateOperator = (op: string) => {
    if (op !== operator) {
      const val = inputRef.current?.props.value?.toString() ?? "";
      const parsed = useDecimal ? parseFloat(val) : parseInt(val, 10);
      dispatcher(
        updatePropertyFilter(containerId, filteringProperty, parsed, op)
      );
    }
  };

  const operatorSelected = (op: string) => operator === op;

  const memoizedFilter = useMemo(
    () => (
      <FilterDecorator
        displayLabel={withAssociatedLabel}
        className={className}
        labelClassName={labelClassName}
        label={label}
        style={style}
      >
        <>
          <InputGroup size="sm">
            <InputGroupAddon addonType="prepend">
              <ButtonDropdown isOpen={dropdownOpen} toggle={toggle}>
                <DropdownToggle className="p-0 m-0 rounded-left text-muted z-index-auto">
                  <FontAwesomeIcon
                    icon={faCaretDown}
                    className="mx-1 text-light"
                  />
                </DropdownToggle>
                <DropdownMenu>
                  <DropdownItem
                    active={operatorSelected(operatorSymbols.eq)}
                    onClick={() => updateOperator(operatorSymbols.eq)}
                  >
                    Equal
                  </DropdownItem>
                  <DropdownItem
                    active={operatorSelected(operatorSymbols.ne)}
                    onClick={() => updateOperator(operatorSymbols.ne)}
                  >
                    Not equal
                  </DropdownItem>
                  <DropdownItem
                    active={operatorSelected(operatorSymbols.gt)}
                    onClick={() => updateOperator(operatorSymbols.gt)}
                  >
                    Greater than
                  </DropdownItem>
                  <DropdownItem
                    active={operatorSelected(operatorSymbols.ge)}
                    onClick={() => updateOperator(operatorSymbols.ge)}
                  >
                    Greater than or equal
                  </DropdownItem>
                  <DropdownItem
                    active={operatorSelected(operatorSymbols.lt)}
                    onClick={() => updateOperator(operatorSymbols.lt)}
                  >
                    Less than
                  </DropdownItem>
                  <DropdownItem
                    active={operatorSelected(operatorSymbols.le)}
                    onClick={() => updateOperator(operatorSymbols.le)}
                  >
                    Less than or equal
                  </DropdownItem>
                </DropdownMenu>
              </ButtonDropdown>
            </InputGroupAddon>
            <InputGroupAddon addonType="prepend">
              <InputGroupText>
                <FontAwesomeIcon
                  icon={faIconByOperator[operator]}
                  style={{ fontSize: ".9em" }}
                />
              </InputGroupText>
            </InputGroupAddon>
            <DebouncedInputField
              inputRef={inputRef}
              filteringProperty={filteringProperty}
              filterValue={filterValue}
              onChange={updateTargetFilter}
              type="number"
              placeholder={placeholder ?? label}
            />
          </InputGroup>
        </>
      </FilterDecorator>
    ),
    [filterValue, dropdownOpen]
  );

  return memoizedFilter;
};

export default NumericFilter;
