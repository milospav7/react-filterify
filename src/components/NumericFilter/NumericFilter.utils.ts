import { ValueTypedObject } from "../../store/types";

const numericOperatorSymbols: ValueTypedObject<string> = {
  eq: "eq",
  ne: "ne",
  gt: "gt",
  ge: "ge",
  lt: "lt",
  le: "le",
};

export { numericOperatorSymbols };
