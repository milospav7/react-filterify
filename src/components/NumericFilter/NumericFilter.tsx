import React, { useState, useMemo, useRef, useCallback } from "react";
import {
  InputGroup,
  DropdownToggle,
  DropdownMenu,
  DropdownItem,
  ButtonDropdown,
  InputGroupText,
  Input,
} from "reactstrap";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCaretDown } from "@fortawesome/free-solid-svg-icons";
import {
  useContainerStyleSchema,
  useFilterActions,
  useFilterState,
} from "../../hooks";
import { DebouncedInputField } from "../shared/DebouncedInputField";
import { BaseFilterProps } from "../../store/interfaces";
import FilterDecorator from "../shared/FilterDecorator";
import RenderIf from "../shared/RenderIf";
import { numericOperatorSymbols } from "./NumericFilter.utils";
import { faIconByOperator } from "../shared/common.utils";

interface IProps extends BaseFilterProps {
  placeholder?: string;
  useDecimal?: boolean;
  multipleOperators?: boolean;
}

const NumericFilter: React.FC<IProps> = ({
  containerId,
  filteringProperty,
  navigationProperty,
  useDecimal,
  placeholder,
  className,
  withAssociatedLabel,
  labelClassName,
  label,
  style,
  multipleOperators,
}) => {
  const [operatorsDropdownOpen, setOperatorsDropdownOpen] = useState(false);
  const toggleDropdown = useCallback(
    () => setOperatorsDropdownOpen(!operatorsDropdownOpen),
    [operatorsDropdownOpen]
  );
  const inputRef = useRef<Input>(null);

  const { updateFilter } = useFilterActions(
    containerId,
    filteringProperty,
    navigationProperty
  );
  const { filterValue, filterOperator } = useFilterState(
    containerId,
    filteringProperty,
    navigationProperty
  );
  const operator = filterOperator ?? numericOperatorSymbols.eq;
  const { styles } = useContainerStyleSchema(containerId);

  const updateTargetFilter = useCallback(
    (value: string | null) => {
      if (value) {
        const parsed = useDecimal ? parseFloat(value) : parseInt(value, 10);
        updateFilter(parsed, { operator });
      } else updateFilter(value, { operator });
    },
    [operator, updateFilter, useDecimal]
  );

  const updateOperator = useCallback(
    (op: string) => {
      if (op !== operator) {
        const val = inputRef.current?.props.value?.toString() ?? "";
        const parsed = useDecimal ? parseFloat(val) : parseInt(val, 10);
        updateFilter(parsed, { operator: op });
      }
    },
    [operator, updateFilter, useDecimal]
  );

  const operatorSelected = useCallback(
    (op: string) => operator === op,
    [operator]
  );

  // TODO: Replace hardcoded items with .map and structured specification
  const filterOperators = useMemo(
    () => (
      <RenderIf condition={!!multipleOperators}>
        <ButtonDropdown isOpen={operatorsDropdownOpen} toggle={toggleDropdown}>
          <DropdownToggle
            className="p-0 m-0 rounded-left text-muted z-index-auto"
            data-testid={`${containerId}-oprs-menu-btn`}
          >
            <FontAwesomeIcon icon={faCaretDown} className="mx-1 text-light" />
          </DropdownToggle>
          <DropdownMenu>
            <DropdownItem
              active={operatorSelected(numericOperatorSymbols.eq)}
              onClick={() => updateOperator(numericOperatorSymbols.eq)}
            >
              Equal
            </DropdownItem>
            <DropdownItem
              active={operatorSelected(numericOperatorSymbols.ne)}
              onClick={() => updateOperator(numericOperatorSymbols.ne)}
            >
              Not equal
            </DropdownItem>
            <DropdownItem
              active={operatorSelected(numericOperatorSymbols.gt)}
              onClick={() => updateOperator(numericOperatorSymbols.gt)}
            >
              Greater than
            </DropdownItem>
            <DropdownItem
              active={operatorSelected(numericOperatorSymbols.ge)}
              onClick={() => updateOperator(numericOperatorSymbols.ge)}
            >
              Greater than or equal
            </DropdownItem>
            <DropdownItem
              active={operatorSelected(numericOperatorSymbols.lt)}
              onClick={() => updateOperator(numericOperatorSymbols.lt)}
            >
              Less than
            </DropdownItem>
            <DropdownItem
              active={operatorSelected(numericOperatorSymbols.le)}
              onClick={() => updateOperator(numericOperatorSymbols.le)}
            >
              Less than or equal
            </DropdownItem>
          </DropdownMenu>
        </ButtonDropdown>
        <InputGroupText>
          <FontAwesomeIcon
            icon={faIconByOperator[operator]}
            className="filter-operator-icon"
          />
        </InputGroupText>
      </RenderIf>
    ),
    [
      containerId,
      operatorsDropdownOpen,
      multipleOperators,
      operator,
      operatorSelected,
      toggleDropdown,
      updateOperator,
    ]
  );

  const memoizedFilter = useMemo(
    () => (
      <FilterDecorator
        displayLabel={withAssociatedLabel}
        className={className}
        labelClassName={labelClassName}
        label={label}
        style={style}
        labelStyle={styles.label}
      >
        <>
          <InputGroup size="sm">
            {filterOperators}
            <DebouncedInputField
              inputRef={inputRef}
              filteringProperty={filteringProperty}
              filterValue={filterValue}
              onChange={updateTargetFilter}
              type="number"
              placeholder={placeholder ?? label}
              style={styles.input}
            />
          </InputGroup>
        </>
      </FilterDecorator>
    ),
    [
      withAssociatedLabel,
      className,
      labelClassName,
      label,
      style,
      styles.label,
      styles.input,
      filterOperators,
      filteringProperty,
      filterValue,
      updateTargetFilter,
      placeholder,
    ]
  );

  return memoizedFilter;
};

export default NumericFilter;
