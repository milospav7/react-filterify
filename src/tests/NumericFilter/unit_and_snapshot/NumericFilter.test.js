import { fireEvent, screen } from "@testing-library/react";
import NumericFilter from "../../../components/NumericFilter/NumericFilter";
import { CONTAINER_IDS } from "../../../store/store";
import { render, rtrCreate } from "../../../test-utils";

/** UNIT TESTING */
test("Numeric filter with label", () => {
  render(
    <NumericFilter
      containerId={CONTAINER_IDS.C2_Test}
      withLabel
      label="Year when joined"
      filteringProperty="YearJoined"
      placeholder="Datetime when joined.."
    />
  );
  expect(
    screen.getByPlaceholderText(/datetime when joined/i)
  ).toBeInTheDocument();
  expect(screen.getByPlaceholderText(/datetime when joined/i)).toHaveValue(
    null
  );
  expect(screen.getByText(/year when joined/i)).toBeInTheDocument();
});

test("Numeric filter without label", () => {
  render(
    <NumericFilter
      containerId={CONTAINER_IDS.C2_Test}
      filteringProperty="YearJoined"
      placeholder="Datetime when joined.."
    />
  );
  expect(screen.queryByText(/year when joined/i)).not.toBeInTheDocument(); // ** getByText will throw error if nothing found
  expect(
    screen.getByPlaceholderText(/datetime when joined/i)
  ).toBeInTheDocument();
  expect(screen.getByPlaceholderText(/datetime when joined/i)).toHaveValue(
    null
  );
});

test("Operators menu will be displayed when toggle button clicked", () => {
  render(
    <NumericFilter
      containerId={CONTAINER_IDS.C2_Test}
      filteringProperty="YearJoined"
      placeholder="Datetime when joined.."
    />
  );
  fireEvent.click(screen.getByTestId(`${CONTAINER_IDS.C2_Test}-oprs-menu-btn`));
  expect(screen.queryByText("Equal")).toBeInTheDocument();
});

// To throw error without container id

/** SNAPSHOT TESTING */
it("Render NumericFilter as expected", () => {
  const filter = rtrCreate(
    <NumericFilter
      containerId={CONTAINER_IDS.C2_Test}
      filteringProperty="YearJoined"
      placeholder="Datetime when joined.."
    />
  ).toJSON();

  expect(filter).toMatchSnapshot();
});
