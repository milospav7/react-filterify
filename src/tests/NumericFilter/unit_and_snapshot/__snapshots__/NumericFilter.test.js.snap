// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render NumericFilter as expected 1`] = `
<div
  className="mb-2"
>
  <div
    className="input-group input-group-sm"
  >
    <input
      className="form-control-sm form-control"
      id="YearJoined-id-dbf"
      name="YearJoined"
      onChange={[Function]}
      placeholder="Datetime when joined.."
      role="textbox"
      style={
        Object {
          "fontSize": undefined,
        }
      }
      type="number"
      value=""
    />
  </div>
</div>
`;
